scipy/sparse/_coo.py

```diff
@@ -651,8 +652,53 @@ def _matmul_vector(self, other):
         return result
 
 
+    def _rmatmul_dispatch(self, other):
+        if isscalarlike(other):
+            return self._mul_scalar(other)
+        else:
+            # Don't use asarray unless we have to
+            try:
+                o_ndim = other.ndim
+                perm = tuple(range(o_ndim)[:-2]) + tuple(range(o_ndim)[-2:][::-1])
+                tr = other.transpose(perm)
+            except AttributeError:
+                o_arr = np.asarray(other)
+                o_ndim = o_arr.ndim
+                perm = tuple(range(o_ndim)[:-2]) + tuple(range(o_ndim)[-2:][::-1])
+                tr = o_arr.transpose(perm)
+            
+            s_ndim = self.ndim
+            perm = tuple(range(s_ndim)[:-2]) + tuple(range(s_ndim)[-2:][::-1])
+            ret = self.transpose(perm)._matmul_dispatch(tr)
+            if ret is NotImplemented:
+                return NotImplemented
+            
+            if s_ndim == 1 or o_ndim == 1:
+                perm = range(ret.ndim)
+            else:
+                perm = tuple(range(ret.ndim)[:-2]) + tuple(range(ret.ndim)[-2:][::-1])
+            return ret.transpose(perm)
+        
+
     def _matmul_dispatch(self, other):
-        if self.ndim < 3:
+        if isscalarlike(other):
+            return self * other
```

Comment:
Arg -- this is causing infinite loops for coo_matrix because `*` is matmul so we come back to `matmul_dispatch`.  Let's use `self.multiply(other)` rather than `self * other`.

That should remove the test failures (at least some of them).

Reply:
Fixed at 53c01967265a76ca75f1afeaaba07dc4fd8ecb18.